### ISSUE TITLE
More than two AND limit for REST query not being honored

### ISSUE THREAD

**Issue by jdockter** [hyperledger/composer#2978]
When using a query with more than two AND limits the second limit is not being honored.

## Context
We have shipment data where the ship to country is Singapore and the ship from country is Mexico.  When trying to limit by both and by status, so three total limits ship to country is not being honored.

query selectShipments {
  description: "Select Shipments By From and To Country and Status"
  statement:
      SELECT com.ibm.customs.Shipment
        WHERE (shipFromAddress.shipFromCountryName==_$fromCountry
        AND shipToAddress.shipToCountryName==_$toCountry
        AND shipmentStatus==_$shipmentStatus)
}

## Expected Behavior
Limit by all three values

## Actual Behavior
Only limits by two

## Steps to Reproduce

1.
2.
3.
4.

## Existing issues

- [ ] [Stack Overflow issues](http://stackoverflow.com/tags/hyperledger-composer)
- [ ] [GitHub Issues](https://github.com/hyperledger/composer/issues)
- [ ] [Rocket Chat history](https://chat.hyperledger.org/channel/composer)

## Your Environment

* Version used:
* Environment name and version (e.g. Chrome 39, node.js 5.4):
* Operating System and version (desktop or mobile):
* Link to your project:

**Comment by fenglianxu** [comment]
@jdockter  Regarding with multiple AND operators,  there is a limitation with PouchDB but should work with CouchDB. 
Can you send the class model for the Shipments so that I can take a look at this issue?

**Comment by jdockter** [comment]
@fenglianxu sent you the cto file

**Comment by fenglianxu** [comment]
@jdockter I have tested two AND operators which is similar to yours with REST server explorer. I can get the results please see the test case:
query findBondByPeriodAndPeriodMultiplierAndMaturity {
    description: "Find all bonds by a specific period _$period and the period multiplier _$multiplier and after a maturity date:_$maturity"
    statement: SELECT org.acme.bond.BondAsset WHERE (bond.paymentFrequency.period == _$period AND bond.paymentFrequency.periodMultiplier <= _$multiplier AND bond.maturity > _$maturity)
}
I can query the results vs couchDB. Slack me when you are available so that we can look at your environment and data set etc.

**Comment by fenglianxu** [comment]
Current status for this defect is: I could not reproduce this issue on our system.  But the test case has returned one more extra data in Jon's( @jdockter) environment. At couch db level, the same test can get the expected result in their environment.  A further investigation is required here.

**Comment by davidkel** [comment]
Someone else is seeing something similar, this was reported on RC
https://chat.hyperledger.org/channel/composer?msg=XXnDyfyM5wdsTvGsS) 
```
Query is not working properly when I have more than 2. The center condition is not working
```

**Comment by varun-raj** [comment]
Even I'm facing the same issue. Is there anyone how has positive results with multiple AND statements?

**Comment by fenglianxu** [comment]
@varun-raj  I  have tested multiple AND queries in via REST server and this works as expected.  Can you provide more details with your query result?
@jdockter Jon's query has got one extra result, I am currently working on this issue.

**Comment by rgallino** [comment]
Hi!
Here I'm facing the same issue too. I've added the query that you described to the bond-network example, and I've added three BondAssets. 

This is the summary of the three BondAssets relevant data to the query:

"ISINCode": "1133",
"maturity": "2017-11-21T13:25:59.617Z",
"periodMultiplier": 2,
"period": "DAY"

"ISINCode": "5983",
"maturity": "2017-10-21T13:28:04.421Z",
"periodMultiplier": 10,
"period": "DAY"

"ISINCode": "2072",
"maturity": "2017-11-29T13:27:28.806Z",
"periodMultiplier": 10,
"period": "DAY"

I ran the query via REST server with the following parameters:

period: DAY
multiplier: 5
maturity: 2016-12-21T13:28:04.421Z

and obtained as a result a list with all the BondAssets, when I should have obtained only one record that satisfy the query parameters. The only one record that I should obtain is 

"ISINCode": "1133",
"maturity": "2017-11-21T13:25:59.617Z",
"periodMultiplier": 2,
"period": "DAY"

which has a periodMultiplier value <= 5. 

I think the problem starts when the query has more than two AND conditions.

**Comment by rgallino** [comment]
Hi there again. We're still fighting with the queries, and we tried intercepting via TCPDUMP the query that arrives to the couchdb when the composer query is executed. The query deployed with de BNA archive is the query that you suggested earlier:

query findBondByPeriodAndPeriodMultiplierAndMaturity {
description: "Find all bonds by a specific period _$period and the period multiplier $multiplier and after a maturity date:$maturity"
statement: SELECT org.acme.bond.BondAsset WHERE (bond.paymentFrequency.period == _$period AND bond.paymentFrequency.periodMultiplier <= _$multiplier AND bond.maturity > _$maturity)
}

And this is the final couchdb Query that we have intercepted with TCPDUMP. 

{
  "limit": 10000,
  "selector": {
    "$and": [
      {
        "chaincodeid": "bond-network"
      },
      {
        "data.$class": "org.acme.bond.BondAsset",
        "data.$registryId": "org.acme.bond.BondAsset",
        "data.$registryType": "Asset",
        "data.bond.paymentFrequency.period": {
          "$eq": "DAY"
        },
        "data.bond.maturity": {
          "$gt": "2017-12-22T17:49:22.333Z"
        }
      }
    ]
  },
  "skip": 0
}

As you can see, it only uses the first and the last conditions of the AND chain. We have been testing with other queries in our BNA network, and in all the cases the conditions between the first and the last condition in the AND chain are allways ingnored.

Also, we've discovered that the issue is in between the v0.15.1 and v0.15.2 version of the composer runtime.

**Comment by fenglianxu** [comment]
The issue has been found in the query compiler and it was caused by the fix to support the pouchDB with multiple combination operators.  
 Pouch DB has limitations to support an explicit combinations of $and and $or operators, a defect has been raised to them. 
There are 2 solution options:
1. to provide the same compiled query for both couch db and pouch db
2. to implement different compiled queries for couch db and pouch db respectively
With option 1,  we need to remove some test cases until the pouch db has fixed the issue, this has less implementation effort. 
With option 2,  a different implementation for pouch db can be provided, but this needs more implementation and test effort.
@sstone1 can you confirm which option we should take?

**Comment by sstone1** [comment]
@fenglianxu please provide the link to the PouchDB issue so we can track it.

With option 2, you say that you can deliver a different implementation that works for PouchDB - is there any reason to think that implementation wouldn't work for CouchDB as well?

**Comment by fenglianxu** [comment]
@sstone1  I have made some changes to accommodate our query compiler to support both couchdb and pouchdb. 
Our query compiler now supports:
1. a combination of multiple **AND** and **OR**
2. a multiple nested **AND** and **OR** operators 
These now bring down the number of pending test cases in the filter. 
The solution to this issue is using the implicit format of the **AND** operator,  but also use the explicit format when the first operator is the **OR**. 
With this solution the query is no longer depending on the  pouch db limitation. I will check in code for you to review.

**Comment by fenglianxu** [comment]
@jdockter per our chat today, I close this defect as there is no need to backport this fix to V0.16.* as you will upgrade your Composer to the V0.17 next stable version.
